test/common/upstream/subset_lb_test.cc

```diff
@@ -1110,6 +1115,50 @@ TEST_P(SubsetLoadBalancerTest, ZoneAwareBalancesSubsetsAfterUpdate) {
   EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][3], lb_->chooseHost(&context));
 }
 
+TEST_F(SubsetLoadBalancerTest, SubsetLogging) {
+  EXPECT_CALL(subset_info_, fallbackPolicy())
+      .WillRepeatedly(Return(envoy::api::v2::Cluster::LbSubsetConfig::DEFAULT_SUBSET));
+
+  const ProtobufWkt::Struct default_subset = makeDefaultSubset({{"version", "default"}});
+  EXPECT_CALL(subset_info_, defaultSubset()).WillRepeatedly(ReturnRef(default_subset));
+
+  std::vector<std::set<std::string>> subset_keys = {{"version"}};
+  EXPECT_CALL(subset_info_, subsetKeys()).WillRepeatedly(ReturnRef(subset_keys));
+
+  std::string expected_log_msgs[] = {
+      "subset lb: creating fallback load balancer for version=\"default\"\n",
+      "subset lb: creating load balancer for version=\"new\"\n",
+      "subset lb: creating load balancer for version=\"default\"\n",
+  };
+
+  std::shared_ptr<Envoy::Filesystem::MockFile> file(new Envoy::Filesystem::MockFile());
+  for (auto& expected : expected_log_msgs) {
+    EXPECT_CALL(*file, write(EndsWith(expected)));
+  }
+
+  NiceMock<Envoy::AccessLog::MockAccessLogManager> log_manager;
+  EXPECT_CALL(log_manager, createAccessLog(_)).WillOnce(Return(file));
+
+  Logger::Registry::getSink()->logToFile("/dev/null", log_manager);
+  const Logger::Logger* upstream_logger = nullptr;
+  for (auto& logger : Logger::Registry::loggers()) {
+    if (logger.name() == "upstream") {
+      upstream_logger = &logger;
+      break;
+    }
+  }
+  ASSERT(upstream_logger != nullptr);
+  upstream_logger->setLevel(spdlog::level::debug);
+
+  init({
+      {"tcp://127.0.0.1:80", {{"version", "new"}}},
+      {"tcp://127.0.0.1:81", {{"version", "default"}}},
+  });
+
+  upstream_logger->setLevel(spdlog::level::err);
```

Comment:
If someone configured a different log level for testing on the CLI with --test_arg, won't this overwrite it after the test runs? I wish there was a better way of doing a test like this, but at minimum should we capture the previous log level so we can reset it?

Reply:
One option would be to just test the print function directly and not the logging. Up to you.